quickwit-doc-mapper/src/tokenizers.rs

```diff
@@ -18,13 +18,106 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use once_cell::sync::Lazy;
-use tantivy::tokenizer::{RawTokenizer, RemoveLongFilter, TextAnalyzer, TokenizerManager};
+use regex::Regex;
+use std::str::CharIndices;
+use tantivy::tokenizer::{
+    BoxTokenStream, RawTokenizer, RemoveLongFilter, TextAnalyzer, Token, TokenStream, Tokenizer,
+    TokenizerManager,
+};
+
+/// Tokenize the text without splitting on ".", "-" and "_" in numbers.
+#[derive(Clone)]
+pub struct LogTokenizer;
+
+pub struct LogTokenStream<'a> {
+    text: &'a str,
+    chars: CharIndices<'a>,
+    token: Token,
+}
+
+impl Tokenizer for LogTokenizer {
+    fn token_stream<'a>(&self, text: &'a str) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(LogTokenStream {
+            text,
+            chars: text.char_indices(),
+            token: Token::default(),
+        })
+    }
+}
+
+impl<'a> LogTokenStream<'a> {
+    fn search_token_end(&mut self) -> usize {
+        (&mut self.chars)
+            .filter(|&(_, ref c)| !c.is_alphanumeric())
+            .map(|(offset, _)| offset)
+            .next()
+            .unwrap_or_else(|| self.text.len())
+    }
+
+    fn search_number_end(&mut self) -> usize {
+        let valid_chars_in_number = Regex::new("[-_.:a-zA-Z]").unwrap();
+        (&mut self.chars)
+            .filter(|&(_, ref c)| {
+                c.is_ascii_whitespace()
+                    || !(c.is_alphanumeric() || valid_chars_in_number.is_match(&c.to_string()))
+            })
+            .map(|(offset, _)| offset)
+            .next()
+            .unwrap_or_else(|| self.text.len())
+    }
+}
+
+impl<'a> TokenStream for LogTokenStream<'a> {
+    fn advance(&mut self) -> bool {
+        self.token.text.clear();
+        self.token.position = self.token.position.wrapping_add(1);
+        while let Some((offset_from, c)) = self.chars.next() {
+            // If the token starts with a number, do not split on punctuation and check if the whole sequence
+            // should be kept (e.g an ip address), if the token does not match, we call search token end
+            if c.is_numeric() {
+                let number = Regex::new(r"\d{1,3}[_.:-]*").unwrap();
+                let mut offset_to = self.search_number_end();
+                if !number.is_match(&self.text[offset_from..offset_to]) {
+                    offset_to = self.search_token_end();
```

Comment:
why calling `self.search_token_end()`? By calling `number.is_match` you did already all the job right? I would look at https://docs.rs/regex/latest/regex/struct.Match.html and see if you can get directly the start and end offset.